# UNICEF- Finding Common Themes

This notebook serves to convert the original dataset into clusters of translated questions with similar themes

In [4]:
!pip install googletrans==4.0.0-rc1

In [36]:
pd.options.mode.chained_assignment = None

In [5]:
#First,the usual imports
import numpy as np
import pandas as pd
import random
from tqdm import tqdm
tqdm.pandas()
import re
from nltk.corpus import stopwords

c:\users\hp\appdata\local\programs\python\python38-32\lib\site-packages\tqdm\std.py:658: FutureWarning: The Panel class is removed from pandas. Accessing it from the top-level namespace will also be removed in the next version
  from pandas import Panel


In [28]:
#Now the aim is to translate all the questions and find the relevant ones
from googletrans import Translator
translator = Translator()

In [7]:
df = pd.read_csv('ureport_sample.csv')

In [8]:
non_eng_polls = df[df['org_language'] != 'en'][['poll_title', 'question_title', 
                                                'org_language', 'poll_category_name']]
#It would make sense to remove duplicates

In [9]:
all_poll_categs = non_eng_polls['poll_category_name'].unique()
print(len(all_poll_categs))
print(all_poll_categs)

299
['Comunicación' 'Inclusión' 'Protección' 'Participación Adolescente'
 'Educación' 'Salud' 'Fechas Importantes' 'Igualdad de Género'
 'Cambio Climático' 'Adolescencia' 'Participación' 'Niñez' 'embarazo'
 'General' 'Ureporteri' 'U-Report Brasil' 'Saúde' 'Educação'
 'ODS 2 - Fome Zero' '+Q' 'Objetivos do Desenvolvimento Sustentável' 'ODS'
 'Mete a colher' 'ODS Geral' 'HIV ' 'Política' 'Migrantes' 'Aprendiz'
 'Esportes' 'Violência Sexual' 'Evasão Escolar' 'Corpo e Gordofobia'
 'ODS 5 - Igualdade de Gênero' 'Corpo' 'Redução' 'Juv e Trabalho'
 'Direitos Reprodutivos e Prevenções' 'Proteção'
 "ODS 14 - Vida debaixo D'Água"
 'ODS 9 - Indústria, Inovação e Infraestrutura'
 'ODS 3 - Saúde e Bem-Estar' 'ODS 1 - Erradicação da Pobreza'
 'Abordagem Policial' 'Acesso à justiça' 'ARMAS' 'DROGAS' 'Rede LGBT'
 'Letalidade Violenta' 'Segurança' 'S4D' 'Opinions' 'Participation'
 'Education' 'Santé' 'Cybercrime' 'Nutrition' 'Hygiène'
 "Droits de l'enfant" 'Général' 'Eau, Assainissement et Hygiène'
 "P

Important Observations:
- Some of these themes are in two languages separated by a slash. 
- Some polls- especially in the Balkan region- simply use 'Polls' or 'U-Report' in the titles. We will have to dig deeper to find the questions within them

There are only 299 categories. We will now create a function to store a dictionary of these 299 mappings (and any changes resulting from the points above). This would prove more efficient than running Google Translate over thousands of lines of code. It would most likely push us past the daily API call limit. 


## Preprocessing for Google Translate

In [10]:
#We are assuming that the language codes are the same in Google Translate and our data
non_eng_polls['org_language'].unique()

array(['es', 'bs', 'pt-br', 'bg', 'fr', 'id', 'ar', 'ro', 'pt', 'my', nan,
       'it', 'sr-rs@latin', 'uz', 'vi', 'uk', 'th'], dtype=object)

Cross-checking against the Google Translate list, we find a few mismatches
https://cloud.google.com/translate/docs/languages


In [11]:
non_eng_polls.loc[non_eng_polls['org_language']=='sr-rs@latin']

poll_title                                     question_title  \
22126  Životna sredina  Da li si informisan/a o stanju životne sredine...   
22127  Životna sredina  Da li si informisan/a o stanju životne sredine...   
22128  Životna sredina  Da li si informisan/a o stanju životne sredine...   
22129  Životna sredina  Da li si informisan/a o stanju životne sredine...   
22130  Životna sredina  Da li si informisan/a o stanju životne sredine...   
...                ...                                                ...   
23221          Nasilje  Kom tipu nasilja nad decom bi najpre trebalo d...   
23222          Nasilje  Na koji način bi deca trebalo da se uključe ka...   
23223          Nasilje  Na koji način bi deca trebalo da se uključe ka...   
23224          Nasilje  Na koji način bi deca trebalo da se uključe ka...   
23225          Nasilje  Na koji način bi deca trebalo da se uključe ka...   

      org_language       poll_category_name  
22126  sr-rs@latin  Zaštita životne sredine  
22127  sr-rs@latin  Zaštita životne sredine  
22128  sr-rs@latin  Zaštita životne sredine  
22129  sr-rs@latin  Zaštita životne sredine  
22130  sr-rs@latin  Zaštita životne sredine  
...            ...                      ...  
23221  sr-rs@latin       Prevencija nasilja  
23222  sr-rs@latin       Prevencija nasilja  
23223  sr-rs@latin       Prevencija nasilja  
23224  sr-rs@latin       Prevencija nasilja  
23225  sr-rs@latin       Prevencija nasilja  

[1100 rows x 4 columns]

A quick check reveals that this is Serbian/Croatian, which is represented in Google translate as 'hr'. 
Brazilian Portuguese can be replaced by Portuguese.

In [12]:
#Completing the first two replacements
non_eng_polls['org_language'].replace('pt-br', 'pt', inplace=True)
non_eng_polls['org_language'].replace('sr-rs@latin', 'hr', inplace=True)

In [13]:
#Exploring the Uzbek case
non_eng_polls.loc[non_eng_polls['org_language']=='uz']

poll_title  \
37217  Karantinda bolalar vaqtini uyda qanday o’tkazy...   
37218  Karantinda bolalar vaqtini uyda qanday o’tkazy...   
37219  Karantinda bolalar vaqtini uyda qanday o’tkazy...   
37220  Karantinda bolalar vaqtini uyda qanday o’tkazy...   
37221  Karantinda bolalar vaqtini uyda qanday o’tkazy...   
...                                                  ...   
38166  Yoshlarning ijtimoiy faolligi / Социальная акт...   
38167  Yoshlarning ijtimoiy faolligi / Социальная акт...   
38168  Yoshlarning ijtimoiy faolligi / Социальная акт...   
38169  Yoshlarning ijtimoiy faolligi / Социальная акт...   
38170  U-Reportda nima haqida muloqot qilmoqchisiz? /...   

                                          question_title org_language  \
37217  2)\tSiz bu bolaga kimsiz? / Кем Вы являетесь д...           uz   
37218  3)\tKarantin tufayli bog'chalar yopilganidan s...           uz   
37219  6)\tBog’chaga borolmaslik, do’stlari, tarbiyac...           uz   
37220  6)\tBog’chaga borolmaslik, do’stlari, tarbiyac...           uz   
37221  6)\tBog’chaga borolmaslik, do’stlari, tarbiyac...           uz   
...                                                  ...          ...   
38166  Tuman/mahallangizni o’zgartirish bo’yicha g’oy...           uz   
38167  Tuman/mahallangizni o’zgartirish bo’yicha g’oy...           uz   
38168  Tuman/mahallangizni o’zgartirish bo’yicha g’oy...           uz   
38169  Tuman/mahallangizni rivojlantirishda ko’ngilli...           uz   
38170  U-Reportda qaysi mavzularda muloqot qilisni is...           uz   

         poll_category_name  
37217  Ta'lim / Образование  
37218  Ta'lim / Образование  
37219  Ta'lim / Образование  
37220  Ta'lim / Образование  
37221  Ta'lim / Образование  
...                     ...  
38166    Yoshlar / Молодежь  
38167    Yoshlar / Молодежь  
38168    Yoshlar / Молодежь  
38169    Yoshlar / Молодежь  
38170    Yoshlar / Молодежь  

[954 rows x 4 columns]

Fortunately, because of its Soviet history, the Uzbek questions carries the Russian translation after a slash. This could come in handy in case the lookup in Uzbek fails. Currently however, it may be a liability by including two languages in one piece of text. So we will remove the Russian that occurs after the slash.  

In [14]:
def keep_preslash(text):
    return text.split('/')[0]

In [15]:
print(keep_preslash('Yoshlar / Молодежь'))

Yoshlar 


In [37]:
non_eng_polls.loc['poll_category_name'] = non_eng_polls['poll_category_name'].apply(lambda x: keep_preslash(x.poll_category_name) if x.org_language=='uz' else x.poll_category_name,
                                                                                    axis=1)

TypeError: <lambda>() got an unexpected keyword argument 'axis'

In [17]:
non_eng_polls.loc[non_eng_polls['org_language']=='uz']['poll_category_name']

37217    Ta'lim / Образование
37218    Ta'lim / Образование
37219    Ta'lim / Образование
37220    Ta'lim / Образование
37221    Ta'lim / Образование
                 ...         
38166      Yoshlar / Молодежь
38167      Yoshlar / Молодежь
38168      Yoshlar / Молодежь
38169      Yoshlar / Молодежь
38170      Yoshlar / Молодежь
Name: poll_category_name, Length: 954, dtype: object

## Translation Function

In [18]:
import time

In [19]:
def translate_to_eng(txt, src_lang):
    """
    takes in text in a non-english language
    returns the english translation
    """
    
    try:
        print('it worked')
        result = translator.translate(txt, 
                     src=src_lang, dest="en")
    #In case the organization's language label doesn't match the question language
    except:
        result = translator.translate(txt, dest="en")
        
    return result.text

In [29]:
def translate_unknown_to_eng(txt):
    """
    takes in text in a non-english language (not specified by user)
    returns the english translation
    """
    
    try:
        result = translator.translate(txt,
                                    dest="en")
        return result.text
    #In case the organization's language label doesn't match the question language
    except:
        return txt

In [21]:
#Testing the function
print(translate_unknown_to_eng('Prevencija nasilja'))

Prevention of violence


In [22]:
print(translate_to_eng('Prevencija nasilja', 'hr'))

it worked
Prevention of violence


In [23]:
time.sleep(10)
print(translate_to_eng('Bonjour', 'fr'))

it worked
Hello


In [24]:
unique_non_eng_polls = non_eng_polls.drop_duplicates(subset=['poll_category_name'])

In [25]:
unique_non_eng_polls

poll_title  \
0      ¿Cuáles son las redes sociales preferidas por ...   
6      ¿Cómo afecta el Aislamiento Preventivo Social ...   
57     ¿Cómo es la situación en el hogar de los adole...   
67     Covid-19: ¿Qué acciones pueden tomar los adole...   
96         Educación a Distancia en contexto de Covid-19   
...                                                  ...   
42397  เราได้จับมือกับสำนักงานกองทุนสนับสนุนการสร้างเ...   
42548  ทุกปียูนิเซฟจะจัดทำรายงานสภาวะเด็กโลก ซึ่งสำรว...   
42572  เราได้ร่วมมือกับกลุ่มการศึกษาเพื่อความเป็นไทอี...   
43915  ภัยพิบัติต่างๆ สร้างความสูญเสียแก่ชุมชนและครอบ...   
44480  เด็กที่อยู่ในครอบครัวยากจน จะขาดโอกาสในหลายๆ ด...   

                                          question_title org_language  \
0                   ¿Cuál es la red social que más usás?           es   
6      Mencione los tres cambios más importantes en l...           es   
57     Durante la cuarentena, ¿quién hace la mayor pa...           es   
67                 ¿Qué tipo de acciones querés conocer?           es   
96     Para quienes se encuentran estudiando actualme...           es   
...                                                  ...          ...   
42397             ปกติน้องๆ เล่นเกมออนไลน์บ่อยแค่ไหนครับ           th   
42548  น้องๆ คิดว่าปกติตัวเองกินอาหารอย่างระวังสุขภาพ...           th   
42572  สิทธิขั้นพื้นฐานในสถานศึกษาด้านไหนที่น้องอยากไ...           th   
43915  น้องๆ คิดว่าตนเองมีความรู้เรื่องการเตรียมความพ...           th   
44480  น้องๆ คิดว่า ในชุมชนที่น้องอยู่ มีคนยากจนหรือไม่?           th   

              poll_category_name  
0                   Comunicación  
6                      Inclusión  
57                    Protección  
67     Participación Adolescente  
96                     Educación  
...                          ...  
42397               Other images  
42548                     สุขภาพ  
42572                  สิทธิเด็ก  
43915           สถานการณ์ฉุกเฉิน  
44480             ความเหลื่อมล้ำ  

[299 rows x 4 columns]

In [30]:
unique_non_eng_polls['poll_category_eng'] = unique_non_eng_polls['poll_category_name'].progress_apply(translate_unknown_to_eng)


 22%|█████████████████▉                                                               | 66/299 [00:13<00:43,  5.37it/s]


 44%|███████████████████████████████████▎                                            | 132/299 [00:26<00:32,  5.08it/s]


 66%|████████████████████████████████████████████████████▉                           | 198/299 [00:39<00:20,  4.90it/s]


 88%|██████████████████████████████████████████████████████████████████████▋         | 264/299 [00:54<00:08,  4.23it/s]


100%|████████████████████████████████████████████████████████████████████████████████| 299/299 [01:02<00:00,  4.78it/s]
<ipython-input-30-e3ab247db721>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  unique_non_eng_polls['poll_category_eng'] = unique_non_eng_polls['poll_category_name'].progress_apply(translate_unknown_to_eng)


In [33]:
unique_non_eng_polls.sort_values(by='poll_category_eng')

poll_title  \
39606                                 Домашнє насильство   
1393                                 Voluntariado UNICEF   
8480     Poll: Apakah citra diri memengaruhi perilakumu?   
1906        Acesso à justiça: para todos ou para alguns?   
19706       Conoscenza e servizi sulla violenza sessuale   
...                                                  ...   
5338                      Día Mundial contra el Mosquito   
42548  ทุกปียูนิเซฟจะจัดทำรายงานสภาวะเด็กโลก ซึ่งสำรว...   
20511   Sondaj despre Obiectivele de Dezvoltare Durabilă   
1042   Un 14% de los nacimientos en Bolivia es de mad...   
1638               Direito ao Esporte Seguro e Inclusivo   

                                          question_title org_language  \
39606  Чи стикалися Ви з домашнім насильством щодо Ва...           uk   
1393   Você está interessado em trabalhar como volunt...           pt   
8480   Seberapa puas atau tidak puaskah kamu terhadap...           id   
1906   Se a escola violar seus direitos, você acessar...           pt   
19706                    Sono...Maschio, Femmina o Altro           it   
...                                                  ...          ...   
5338   1. ¿A cuántas personas mata el mosquito/zancud...           es   
42548  น้องๆ คิดว่าปกติตัวเองกินอาหารอย่างระวังสุขภาพ...           th   
20511  Ai auzit până acum despre Obiectivele de Dezvo...           ro   
1042         ¿Conoces casos de adolescentes embarazadas?           es   
1638   A escola possui equipamentos para prática espo...           pt   

                       poll_category_name                   poll_category_eng  
39606  #EndViolence - Протидія насильству  #EndViolence - Countering violence  
1393                                   +Q                                  +Q  
8480                                 ADAP                                ADAP  
1906                     Acesso à justiça                   Access to justice  
19706                  Accesso ai servizi                  Access to services  
...                                   ...                                 ...  
5338                                 Zika                                Zika  
42548                              สุขภาพ                              health  
20511                             Sondaje                               polls  
1042                             embarazo                           pregnancy  
1638                             Esportes                              sports  

[299 rows x 5 columns]

This sorting behaviour seems odd. Not clear why health, polls, and sports are at the bottom, after Z (Zika)

In [35]:
sorted(unique_non_eng_polls['poll_category_eng'].unique())

['#EndViolence - Countering violence',
 '+Q',
 'ADAP',
 'Access to justice',
 'Access to services',
 'Adolescence',
 'Adolescent Health',
 'Adolescent Participation',
 'Adolescent participation',
 'Advocacy 4',
 'Advocacy and Participation',
 'Apprentice',
 'Art',
 'Aspirations',
 'Average',
 'Black Gold',
 'Body',
 'Body and Gordophobia',
 'CHILD FRIENDLY CITIES',
 'CHILD PROTECTION',
 'CORONAVIRUS 2019 - (COVID-19)',
 'COVID19',
 'Career',
 'Cheers',
 'Child Protection',
 'Child Protection - Identity Card',
 'Child Rights',
 'Child marriage',
 'Child protection',
 'Childhood',
 'Childhood and adolescence',
 'Children on the Move',
 "Children's rights",
 "Children's rights & monitoring",
 'Citizenship',
 'Climate Change',
 'Climate Change 2',
 'Climate and Environment',
 'Climate change',
 'Commitment / solidarity',
 'Common',
 'Communication',
 'Communication Externe',
 'Consumption',
 'Contingency',
 'Corona Virus Response',
 'Culture',
 'Cyber Bullying',
 'Cybercrime',
 'Cyclone ID

From this initial look, it seems like we have a few themes that are showing up across countries.
- We might need to remove the upper case letters in some cses
- The Uzbek terms, as expected, had two languages in the original title. So only the Russian has been translated
- Some polls are simply titled 'Polls'. We will need a separate sub-category for them

